pkg/proctree/tree/tree.go

```diff
@@ -0,0 +1,190 @@
+	processes map[uint32]*nodes.Process // map[murmur3 hash of process] -> process node
+	mutex     *sync.RWMutex             // synchronize the tree
+}
+
+// String returns a string representation of the tree (for debugging purposes)
+func (t *Tree) String() string {
+	wideOutput := false
+
+	buffer := bytes.NewBufferString("")
+
+	// Use tablewriter to print the tree in a table
+	newTable := func() *tablewriter.Table {
+		table := tablewriter.NewWriter(buffer)
+		table.SetHeader([]string{"PPID", "PID", "PCMD", "CMD"})
+		table.SetColMinWidth(0, 8)
+		table.SetColMinWidth(1, 8)
+		table.SetColMinWidth(2, 8)
+		table.SetColMinWidth(3, 8)
+		table.SetAutoWrapText(!wideOutput)
+		table.SetRowLine(!wideOutput)
+		table.SetAutoFormatHeaders(true)
+		table.SetHeaderAlignment(tablewriter.ALIGN_LEFT)
+		table.SetAlignment(tablewriter.ALIGN_LEFT)
+		table.SetHeaderLine(true)
+		table.SetBorder(true)
+		return table
+	}
+
+	table := newTable()
+
+	var lines [][]string
+
+	for processHash, process := range t.processes {
+		pid := fmt.Sprintf("%v", process.GetInfo().Pid)
+		parentPid := fmt.Sprintf("%v", t.GetParentPID(processHash))
+		cmd := process.GetCommand().Name
+		parentCmd := t.GetParent(processHash).GetCommand().Name
+		lines = append(lines, []string{parentPid, pid, parentCmd, cmd})
+	}
+
+	sort.Slice(lines, func(i, j int) bool {
+		firstOne, _ := strconv.Atoi(lines[i][0])
+		firstTwo, _ := strconv.Atoi(lines[j][0])
+		secondOne, _ := strconv.Atoi(lines[i][1])
+		secondTwo, _ := strconv.Atoi(lines[j][1])
+		if firstOne != firstTwo {
+			return firstOne < firstTwo
+		}
+		return secondOne < secondTwo
+	})
+
+	for _, line := range lines {
+		table.Append(line)
+	}
+	table.Render()
+
+	return buffer.String()
+}
+
+// NewTree returns a new tree
+func NewTree() *Tree {
+	return &Tree{
+		processes: make(map[uint32]*nodes.Process),
+		mutex:     &sync.RWMutex{},
+	}
+}
+
+// GetOrCreateProcess returns the process indexed by given hash, or creates a new one if not found.
+func (t *Tree) GetOrCreateProcess(hash uint32) *nodes.Process {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	if _, ok := t.processes[hash]; !ok {
+		t.processes[hash] = nodes.NewProcess(hash)
+	}
+
+	return t.processes[hash]
+}
+
+// GetProcess returns the process indexed by given hash.
+func (t *Tree) GetProcess(hash uint32) (*nodes.Process, bool) {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+	process, ok := t.processes[hash]
+	return process, ok
+}
+
+// GetParent returns the process indexed by given hash parent.
+func (t *Tree) GetParent(hash uint32) *nodes.Process {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+
+	process, ok := t.processes[hash]
+	if !ok {
+		return nodes.NewProcess(0) // empty process node
+	}
+	parentHash := process.GetParentHash()
+	if parentHash == 0 {
+		return nodes.NewProcess(0) // empty process node
+	}
+	parent, ok := t.processes[parentHash]
+	if !ok {
+		return nodes.NewProcess(0) // empty process node
+	}
+
+	return parent
```

Comment:
You are totally right. I'm thinking on keeping the "TASK" concept (instead of process and/or thread) and marking if the TASK is LWP. If its LWP then its parent will be the process leader group parent (just like you said in other comments).

Reply:
As I wrote in other comments, I think like @yanivagman that in user space we need the process/threads approach and not the task approach. Its just much more usable for security checks, which are the goal of this feature.